multi/stm32l4-multi/rng.c

```diff
@@ -0,0 +1,155 @@
+/*
+ * Phoenix-RTOS
+ *
+ * STM32L4 RNG driver
+ *
+ * Copyright 2024 Phoenix Systems
+ * Author: Aleksander Kaminski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <board_config.h>
+#include <stdint.h>
+#include <errno.h>
+#include <string.h>
+#include <sys/threads.h>
+#include <sys/platform.h>
+#include <sys/interrupt.h>
+#include <sys/types.h>
+#include <sys/pwman.h>
+#include <phoenix/arch/stm32l4.h>
+
+#include "common.h"
+
+#define RNG_CR 0
+#define RNG_SR 1
+#define RNG_DR 2
+
+
+static struct {
+	volatile uint32_t *base;
+	handle_t lock;
+} rng_common;
+
+
+static int rng_readDR(uint32_t *val)
+{
+	int ret = -EAGAIN;
+	uint32_t sr = *(rng_common.base + RNG_SR);
+
+	if ((sr & 1u) != 0) {
+		*val = *(rng_common.base + RNG_DR);
+		ret = (*val != 0) ? 0 : -EIO;
+	}
+
+	/* Has error been detected? */
+	if ((sr & (3u << 5)) != 0) {
+		/* Mark data as faulty */
+		ret = -EIO;
+
+		/* Clear flags */
+		*(rng_common.base + RNG_SR) &= ~(sr & (3u << 5));
+		dataBarier();
+	}
+
+	return ret;
+}
+
+
+static void rng_enable(void)
+{
+	platformctl_t pctl = {
+		.action = pctl_set,
+		.type = pctl_devclk,
+		.devclk = {
+			.dev = pctl_hsi48,
+			.state = 1 }
+	};
+
+	/* Enable HSI48 */
+	platformctl(&pctl);
+
+	/* Enable RNG */
+	*(rng_common.base + RNG_CR) |= 1u << 2;
+}
+
+
+static void rng_disable(void)
+{
+	/* Disable RNG */
+	*(rng_common.base + RNG_CR) &= ~(1u << 2);
+
+	platformctl_t pctl = {
+		.action = pctl_set,
+		.type = pctl_devclk,
+		.devclk = {
+			.dev = pctl_hsi48,
+			.state = 0 }
+	};
+
+	/* Disable HSI48 */
+	platformctl(&pctl);
+}
+
+
+ssize_t rng_read(uint8_t *val, size_t len)
+{
+	mutexLock(rng_common.lock);
+
+	rng_enable();
+
+	size_t pos = 0;
+	while (pos < len) {
+		uint32_t t;
+		dataBarier();
+		if (rng_readDR(&t) < 0) {
```

Comment:
According to RM, after noise souce error, you should restart RNG with RNGEN clear and set:

RM0351 Rev 9
>  In order to fully recover from a seed error application 
must clear the SEIS bit by writing it to
> “0”, then clear and set the RNGEN bit to reinitialize and restart the RNG.